packages/yarnpkg-fslib/sources/algorithms/opendir.ts

```diff
@@ -0,0 +1,82 @@
+import {NoParamCallback}     from 'fs';
+
+import {Dir, Dirent, FakeFS} from '../FakeFS';
+import * as errors           from '../errors';
+import {Filename, Path}      from '../path';
+
+export type CustomDirOptions = {
+  onClose?: () => void;
+};
+
+export class CustomDir<P extends Path> implements Dir<P> {
+  constructor(
+    public readonly path: P,
+    private readonly nextDirent: () => Dirent | null,
+    private readonly opts: CustomDirOptions = {},
+  ) {}
+
+  public closed: boolean = false;
+
+  throwIfClosed() {
+    if (this.closed) {
+      throw errors.ERR_DIR_CLOSED();
+    }
+  }
+
+  async * [Symbol.asyncIterator]() {
+    let dirent: Dirent | null;
+    // eslint-disable-next-line no-cond-assign
+    while ((dirent = await this.read()) !== null) {
+      yield dirent;
+    }
+  }
```

Comment:
Looking at the node implementation this should close the `Dir` once complete
https://github.com/nodejs/node/blob/b5490d02d41af17baf4cf25a82aac110cba40ecb/lib/internal/fs/dir.js#L199-L211

Reply:
Thanks, missed that!